File: sistema_java/codigos/oficinapoo/Agenda.java
package oficinapoo;

import java.util.List;
import java.util.LinkedList;
import java.util.Collections;

public class Agenda 
{
    public List<ItemAgenda> itensDaAgenda;

    //construtores
    public Agenda()
    {
        itensDaAgenda = new LinkedList<ItemAgenda>();
    }
    
    public Agenda(Agenda agenda)
    {
        itensDaAgenda = new LinkedList<ItemAgenda>();
        
        for(ItemAgenda it : agenda.itensDaAgenda)
        {
            itensDaAgenda.add(new ItemAgenda(it));
        }     
    }
    
    
    //ordena a agenda pela data
    public void ordenaAgenda()
    {
        Collections.sort(itensDaAgenda);                                                             
    }
     
    
    public void adicionarItemAgenda(ItemAgenda item)//adiciona um Item e ordena a Agenda
    {
	//inserir novo elemento
	ItemAgenda copia = new ItemAgenda(item);
        itensDaAgenda.add(copia);
        ordenaAgenda();
    }

    public boolean removerItemAgenda(int posicao)//dada uma posicao remove um item da agenda
    {
        //verifica se posicao eh valida
	if (itensDaAgenda.isEmpty() || posicao >  itensDaAgenda.size() || posicao < 1)
	{
            return false;
	}
	
        itensDaAgenda.remove(posicao-1);
        
	return true;
    }

    public ItemAgenda retornarItemAgenda(int posicao)//retorna um item da agenda
    {
        //se a posicao for incorreta retorna o ultimmo
        if(posicao > itensDaAgenda.size() || posicao < 1)
        {
            return itensDaAgenda.get(itensDaAgenda.size()-1);
        }
        
	return itensDaAgenda.get(posicao-1);
    }

    //imprime a agenda
    public String toString()
    {
        String retorno = "";
        if (itensDaAgenda.isEmpty())//se nao tiver itens na agenda
	{
		retorno = "Agenda vazia";
	}
        else
        {
            retorno = retorno+"Agenda\n";
            int i = 1;
            for(ItemAgenda item: itensDaAgenda)
            {
                retorno = retorno+"Item " + i +":\n" +item.toString() + "\n";
                i++;
            }
        }
        
        return retorno;
    }  
}

File: sistema_cpluplus/codigos/Agenda.cpp
#include "Agenda.h"
#include <iostream>//entrada e saida


using namespace std;

//construtor
Agenda::Agenda() {}


//construtor
Agenda::Agenda(Agenda& agenda)
{
	//inserir novos elementos
	for (auto it = agenda.itensDaAgenda.begin(); it != agenda.itensDaAgenda.end(); it++)
	{
		this->itensDaAgenda.push_back(new ItemAgenda(**it));
	}

}

//destrutor

Agenda:: ~Agenda()
{
	for (auto item : this->itensDaAgenda)
	{
		if (item != nullptr)
		{
			delete item;
		}
	}
}

void Agenda::adicionarItemAgenda(ItemAgenda item)//adiciona um Item na agenda de maneira ordenada pela data
{
	//inserir novos elementos
	ItemAgenda* copia = new ItemAgenda(item);

	list < ItemAgenda*> ::iterator it = this->itensDaAgenda.begin();

	for (; it != this->itensDaAgenda.end(); it++)
	{
		if ((**it).data > (*copia).data)//se o ano a inserir for menor ja pode inserir
		{
			break;
		}

		else if ((*copia).data > (**it).data)//se for maior, pode continuar a iterar ate achar o ano
		{
			continue;
		}

		else//eh igual
		{
			break;
		}

	}

	this->itensDaAgenda.insert(it, copia);
}

bool Agenda::removerItemAgenda(int posicao)//dada uma posicao remove um item da agenda
{
	if (itensDaAgenda.empty() || posicao >  itensDaAgenda.size() || posicao < 1)
	{
		return false;
	}
	int i = 0;
	for (auto it =itensDaAgenda.begin(); it != itensDaAgenda.end(); it++)
	{
		if (i == posicao - 1)
		{
			delete* it;
			itensDaAgenda.erase(it);
			break;
		}
		i++;
	}

	return true;
}

ItemAgenda& Agenda::retornarItemAgenda(int posicao)//retorna um item da agenda
{
	int i = 0;

	auto it = itensDaAgenda.begin();
	for (; it != itensDaAgenda.end(); it++)
	{
		if (i == posicao - 1)
		{

			return **it;
		}
		i++;
	}

	return **it;///retorna ultimo caso nao encontre
}

//sobrecarga do operador de atribuicao '=' 

Agenda& Agenda:: operator=(Agenda& agenda)
{
	if (this == &agenda)
	{
		return *this;
	}


	//desalocar lista
	for (auto desalocar : this->itensDaAgenda)
	{
		if (desalocar != nullptr)
		{
			delete desalocar;
		}
	}

	this->itensDaAgenda.clear();//limpar a lista 

	//inserir novos elementos
	for (auto it = agenda.itensDaAgenda.begin(); it != agenda.itensDaAgenda.end(); it++)
	{
		this->itensDaAgenda.push_back(new ItemAgenda(**it));
	}

	return *this;
}


//sobrecarga do operador 'cout'
ostream& operator << (ostream& out, Agenda& agenda)
{
	if (agenda.itensDaAgenda.empty())//se nao tiver itens na agenda
	{
		out << "Agenda vazia";
	}
	else//caso contrario imprime os itens
	{
		out << "Agenda:\n ";
		int i = 1;
		for (auto it = agenda.itensDaAgenda.begin(); it != agenda.itensDaAgenda.end(); it++)
		{
			out << "Item " << i << ": " << "\n" << **it << "\n";
			i++;
		}
	}

	return out;
}